![consumer_behaviour.png](consumer_behaviour.png)

# DATA-DRIVEN INSIGHTS ON USER BEHAVIOR AND AD CLICKS
***
## BUSINESS UNDERSTANDING
***

### Business Overview
In the digital age, advertising is a critical component of customer engagement and conversion. However, not all ads perform equally well, and understanding what drives user engagement is key to optimizing ad campaigns. By analyzing user demographics, behavior, and ad characteristics, businesses can tailor their strategies to maximize click-through rates (CTR) and improve return on investment (ROI).

Online advertising is a colossal business worth more than [$50 billion](https://www.researchgate.net/publication/336666758_Machine_Learning_Based_Ad-click_prediction_system), with advertisers increasingly focusing on targeted ad strategies. As the industry grows, it becomes essential to continuously measure the effectiveness of ads and refine strategies accordingly. One of the key metrics for measuring ad performance is the ad click, which indicates how often users interact with online advertisements. A higher number of ad clicks typically signifies a stronger interest in the advertised product, service, or content, making it a valuable indicator of user engagement and potential conversion.

To stay ahead of the competition and optimize ad spending, businesses like Hubspot are leveraging advanced analytics to gain deeper insights into what drives ad engagement. Techniques like A/B testing allow businesses to experiment with different ad variations (e.g., copy, images, targeting) to determine which ones produce the best results. By testing multiple versions of ads with different audience segments, businesses can identify the most effective strategies and apply them at scale, thus improving engagement rates and maximizing ad spend efficiency.

Additionally, machine learning models can be implemented to predict user behavior based on historical ad interactions and demographic data, enabling businesses to target the right audience with the most relevant ads. This combination of A/B testing and machine learning can significantly enhance ad performance, providing data-driven strategies to refine advertising efforts over time.

### Problem Statement
Affiliate Marketing Company, Hubspot seeks to further invest significant resources in online advertising to drive user engagement, awareness, and conversions. However, the challenge often lies in determining which elements of an advertisement – such as content, targeting, timing, and format – are most effective in driving user interaction

### Main Objectives
Our main objective is to determine factors influencing user interaction with online ads and identify actionable insights to optimize ad performance.

#### Specific Objectives
1. Correlation between higher ad clicks and demographic/behavioral factors
2. Investigate the variations in ad content that affect engagement
3. validation of these insights through AB testing to recommend data_driven strategies


## DATA UNDERSTANDING
***

The data utilized for this project was gotten from the [link](https://statso.io/wp-content/uploads/2023/01/ctr.zip)

The dataset contains 10,000 rows and 10 columns with the column descriptions of the data as below. The data comprises of numerical, categorical and datetime data.

| **Feature Name**            | **Description**                                                 | **Type**           | **Examples**               |
|-----------------------------|---------------------------------------------------------------|--------------------|----------------------------|
| Daily Time Spent on Site    | Time spent by a user on the site daily (in minutes)           | Continuous Numeric | 68.95, 80.23               |
| Age                         | Age of the user                                               | Continuous Numeric | 35, 23                     |
| Area Income                 | Average income of the user's geographical area (in dollars)  | Continuous Numeric | 55000, 72000               |
| Daily Internet Usage        | Daily usage of the internet by the user (in minutes)         | Continuous Numeric | 120.5, 96.8                |
| Ad Topic Line               | Topic headline of the ad viewed                              | Categorical Text   | "Top Ad Offer", "Great Deal" |
| City                        | City where the user resides                                   | Categorical Text   | "New York", "San Francisco" |
| Gender                      | Gender of the user                                            | Categorical        | Male, Female               |
| Country                     | Country of the user                                           | Categorical Text   | "USA", "India"             |
| Timestamp                   | Date and time of the ad interaction                          | DateTime           | 2024-10-31 14:53:00        |
| Clicked on Ad               | Whether the user clicked on the ad (1 = Yes, 0 = No)         | Binary Categorical | 1, 0                       |


## DATA PREPARATION
***

The following steps in summary shall be followed in the data preparation stage in preparation for Modeling in later stages

 **1. Data Loading**      
- Load the Datasets  
- Inspect the Data

 **2. Data Cleaning**
- Validity Check
- Consistency Check
- Uniformity Check
- Completeness Check
  
 **3. Exploratory Data Analysis**
- Understand Data Distribution     
- Identify Relationships - Univariate and Bivariate Analysis
- Handle High Cardinality Columns

***
#### 1. DATA LOADING 
***
The following was carried out 
1. Loading the Datasets  
2. Inspecting the Data


In [7]:
import numpy as np
import pandas as pd
from classes import *

# Libraries for visualizations
import folium
import seaborn as sns
import plotly.express as px
from IPython.display import display, HTML
import matplotlib.pyplot as plt
%matplotlib inline

# libraries for Model Preprocessing
from sklearn.model_selection import train_test_split
from sklearn.preprocessing   import StandardScaler, OneHotEncoder, LabelEncoder, MinMaxScaler
from imblearn.over_sampling  import SMOTE


In [8]:
# Instantiate the DataLoader class
loader= DataLoader()

# Reading the csv file
data=loader.read_data("Clicked_On_AD.csv")


# Instantiate the DataInfo class
summary= DataInfo()

# Summary view of the dataset information
summary.info(data)

Shape of the dataset : (10000, 10) 

Column Names
Index(['Daily Time Spent on Site', 'Age', 'Area Income',
       'Daily Internet Usage', 'Ad Topic Line', 'City', 'Gender', 'Country',
       'Timestamp', 'Clicked on Ad'],
      dtype='object') 
 

Data Summary
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 10 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Daily Time Spent on Site  10000 non-null  float64
 1   Age                       10000 non-null  int64  
 2   Area Income               10000 non-null  float64
 3   Daily Internet Usage      10000 non-null  float64
 4   Ad Topic Line             10000 non-null  object 
 5   City                      10000 non-null  object 
 6   Gender                    10000 non-null  object 
 7   Country                   10000 non-null  object 
 8   Timestamp                 10000 non-null  object 
 9   Clicked on Ad        

Daily Time Spent on Site           Age   Area Income  \
count              10000.000000  10000.000000  10000.000000   
mean                  61.660757     35.940100  53840.047721   
std                   15.704142      8.572973  13343.708718   
min                   32.600000     19.000000  13996.500000   
25%                   48.860000     29.000000  44052.302500   
50%                   59.590000     35.000000  56180.930000   
75%                   76.580000     42.000000  61840.260000   
max                   90.970000     60.000000  79332.330000   

       Daily Internet Usage  Clicked on Ad  
count          10000.000000   10000.000000  
mean             177.759831       0.491700  
std               40.820951       0.499956  
min              105.220000       0.000000  
25%              140.150000       0.000000  
50%              178.920000       0.000000  
75%              212.670000       1.000000  
max              269.960000       1.000000

Dataset Overview


Daily Time Spent on Site  Age  Area Income  Daily Internet Usage  \
0                     62.26   32     69481.85                172.83   
1                     41.73   31     61840.26                207.17   
2                     44.40   30     57877.15                172.83   
3                     59.88   28     56180.93                207.17   
4                     49.21   30     54324.73                201.58   

                         Ad Topic Line             City  Gender  \
0      Decentralized real-time circuit         Lisafort    Male   
1       Optional full-range projection  West Angelabury    Male   
2  Total 5thgeneration standardization        Reyesfurt  Female   
3          Balanced empowering success      New Michael  Female   
4  Total 5thgeneration standardization     West Richard  Female   

                        Country        Timestamp  Clicked on Ad  
0  Svalbard & Jan Mayen Islands   6/9/2016 21:43              0  
1                     Singapore  1/16/2016 17:56              0  
2                    Guadeloupe  6/29/2016 10:50              0  
3                        Zambia  6/21/2016 14:32              0  
4                         Qatar  7/21/2016 10:54              1

***
#### 2. DATA CLEANING 
***

**2.1 Consistency Check**
 We shall check for duplicates in the rows and drop any duplicated rows

In [10]:
# Instansitate the DataCheck class
cleaning = DataChecks(data)

# Checking for duplicates
cleaning.check_duplicates()

*****************************************************************
Total Number of Rows Duplicated 341
*****************************************************************




**Observation**
- There are a total of 341 duplicated rows
- Upon confirmation of the duplicates these rows have been dropped to reduce noise

In [24]:
data["Ad Topic Line"].value_counts()

Ad Topic Line
Cloned explicit middleware                 344
Streamlined homogeneous analyzer           212
Business-focused transitional solution     207
Sharable reciprocal project                187
Intuitive exuding service-desk             173
                                          ... 
Ergonomic neutral portal                     1
Implemented uniform synergy                  1
Visionary multi-tasking alliance             1
Inverse zero tolerance customer loyalty      1
Reactive interactive protocol                1
Name: count, Length: 559, dtype: int64

In [16]:
data["City"].value_counts()

City
Hubbardmouth       336
West Brandonton    275
Lisafort           261
Wintersfort        230
Stewartbury        207
                  ... 
Turnerview           1
North Tiffany        1
North Kristine       1
East Troyhaven       1
Davidview            1
Name: count, Length: 521, dtype: int64

In [20]:
data[data['City']=='Hubbardmouth']

Daily Time Spent on Site  Age  Area Income  Daily Internet Usage  \
52                       52.84   42     60514.05                199.43   
82                       82.07   55     57868.44                187.76   
220                      68.88   42     60514.05                236.87   
225                      88.04   41     64828.00                236.87   
243                      46.61   41     55121.65                126.39   
...                        ...  ...          ...                   ...   
9785                     83.71   42     60514.05                199.43   
9801                     89.05   42     57868.44                114.69   
9827                     46.61   23     25739.09                236.87   
9957                     82.07   55     62784.85                187.76   
9958                     46.14   23     61617.98                172.81   

                                     Ad Topic Line          City  Gender  \
52                Virtual 5thgeneration neural-net  Hubbardmouth    Male   
82                   Inverse high-level capability  Hubbardmouth    Male   
220                  Innovative interactive portal  Hubbardmouth  Female   
225            User-friendly grid-enabled analyzer  Hubbardmouth  Female   
243                Synergistic reciprocal attitude  Hubbardmouth  Female   
...                                            ...           ...     ...   
9785      Face-to-face mission-critical definition  Hubbardmouth    Male   
9801                 Robust logistical utilization  Hubbardmouth    Male   
9827            Managed disintermediate capability  Hubbardmouth  Female   
9957             Pre-emptive transitional protocol  Hubbardmouth  Female   
9958  Future-proofed coherent budgetary management  Hubbardmouth    Male   

                       Country        Timestamp  Clicked on Ad  
52              Czech Republic  2/15/2016 16:52              1  
82    Northern Mariana Islands   4/21/2016 9:30              1  
220             Czech Republic   4/6/2016 11:24              1  
225                       Cuba  5/19/2016 14:30              1  
243                   Honduras   5/27/2016 8:53              1  
...                        ...              ...            ...  
9785            Czech Republic  5/19/2016 14:30              1  
9801            Czech Republic   4/6/2016 11:24              1  
9827                     China   4/27/2016 4:28              1  
9957                     Kenya  1/15/2016 22:49              1  
9958                   Denmark    3/9/2016 3:41              0  

[336 rows x 10 columns]

In [ ]:
Null Hypothesis (H₀): There is no difference between groups (e.g., gender or time spent on-site) in terms of ad clicks.
Alternative Hypothesis (H₁): There is a statistically significant difference between groups.

In [ ]:
Segment the Data
Create groups for comparison:

Group A: Users who meet one criterion (e.g., low daily time spent on the site).
Group B: Users who meet the other criterion (e.g., high daily time spent on the site).
For example:

Split users into age groups (e.g., below 30 vs. 30 and above).
Compare users by daily time spent on-site (e.g., <60 minutes vs. ≥60 minutes).


In [ ]:
3. Choose a Metric
Define the metric you want to measure, such as:

Click Rate: Percentage of users who clicked on ads.
Time Spent: Average time spent on-site for each group.

In [ ]:
erform the Test
Perform statistical tests to compare groups:

T-Test or Z-Test: For continuous variables like daily time spent on-site.
Chi-Square Test: For categorical variables like gender or city.
Mann-Whitney U Test: If the data is not normally distributed.

In [ ]:
 Analyze the Results
Check the p-value to determine significance (e.g., p < 0.05).
If the null hypothesis is rejected, conclude that there is a statistically significant difference.

In [ ]:
Example Analysis: Daily Time Spent on Site and Ad Clicks
If you want to test whether users who spend more time on the site are more likely to click ads:

Divide users into two groups:
Group A: Users with daily time ≤ the median.
Group B: Users with daily time > the median.
Calculate the click rates for both groups.
Perform a statistical test to compare the two groups.


In [5]:
data.columns

Index(['Daily Time Spent on Site', 'Age', 'Area Income',
       'Daily Internet Usage', 'Ad Topic Line', 'City', 'Gender', 'Country',
       'Timestamp', 'Clicked on Ad'],
      dtype='object')